src/read.ts

```diff
@@ -0,0 +1,121 @@
+import Mute from 'mute-stream'
+import { Completer, AsyncCompleter, createInterface, ReadLineOptions } from 'readline'
+
+export namespace Read {
+  export interface Options<T extends string | number = string> {
+    default?: T
+    input?: ReadLineOptions['input'] & {
+      isTTY?: boolean
+    }
+    output?: ReadLineOptions['output'] & {
+      isTTY?: boolean
+    }
+    prompt?: string
+    silent?: boolean
+    timeout?: number
+    edit?: boolean
+    terminal?: boolean
+    replace?: string,
+    completer?: Completer | AsyncCompleter,
+  }
+}
+
+export default async function read<T extends string | number = string> ({
+  default: def,
+  input = process.stdin,
+  output = process.stdout,
+  completer,
+  prompt = '',
+  silent,
+  timeout,
+  edit,
+  terminal,
+  replace,
+}: Read.Options<T>): Promise<T | string> {
+  if (
+    typeof def !== 'undefined' &&
+    typeof def !== 'string' &&
+    typeof def !== 'number'
+  ) {
+    throw new Error('default value must be string or number')
+  }
+
+  let editDef = false
+  prompt = prompt.trim() + ' '
+  terminal = !!(terminal || output.isTTY)
+
+  if (def) {
+    if (silent) {
+      prompt += '(<default hidden>) '
+      /* c8 ignore start */
+    } else if (edit) {
+      editDef = true
+      /* c8 ignore stop */
+    } else {
+      prompt += '(' + def + ') '
+    }
+  }
+
+  const m = new Mute({ replace, prompt })
+  m.pipe(output, { end: false })
+  output = m
+
+  return new Promise<string | T>((resolve, reject) => {
+    const rl = createInterface({ input, output, terminal, completer })
+    /* c8 ignore start */
+    const timer = timeout && setTimeout(() => onError(new Error('timed out')), timeout)
+    /* c8 ignore stop */
+
+    m.unmute()
+    rl.setPrompt(prompt)
+    rl.prompt()
+
+    if (silent) {
+      m.mute()
+      /* c8 ignore start */
+    } else if (editDef) {
+      // @ts-ignore
```

Comment:
Matter of preference/style, but I personally like to avoid having lots of sequential `//@ts-ignore` lines, and instead declare an alias as the thing I intend/expect it to be. Eg:

```ts
// might need "as unknown" in there if TS complains
const rlEdit = rl as typeof rl & {
  line: string,
  cursor: number,
  _refreshLine: () => void,
}
rlEdit.line = def
rlEdit.cursor = def.length
rlEdit._refreshLine()
```

The benefit of this is that the reason for the override is captured, along with the interface that you are expecting it to be held to.

Reply:
I like that pattern. Pretty sure it caught a subtle bug too where default is set to `0`.